design/build-clusters.md

```diff
@@ -0,0 +1,160 @@
+| --- | --- | --- | --- | --- | --- |
+| `id` | integer | no | yes | yes | |
+| `name` | text (100) | no | no | yes | |
+| `scmOrganizations` | text(500) | yes | no | no | |
+| `scmContext` | text(200) | no | yes | no | |
+| `isActive` | boolean | no | no | no | *0-false or 1-true* |
+| `managedBy` | text(50) | no | no | no | cluster managed by *screwdriver or external* |
+| `managedByEmail` | text(100) | yes | no | no | cluster admin email for communications |
+
+Unique constraint: `name + isActive` 
+
+#### Sample record
+
+| id | name | scmContext | scmOrganizations | isActive | managedBy | managedByEmail 
+| --- | --- | --- | --- | --- | --- | --- | --- | --- | 
+| 1 | gq1 | github:git.ouroath.com | null | 0 | screwdriver | sd@oath.com
+| 2 | bf1 | github:git.ouroath.com | null | 1 | screwdriver | sd@oath.com
+| 3 | identity | github:git.ouroath.com | identity_org1, identity_org2 | 1 | external | identity@oath.com
+| 4 | identity | github:git.ouroath.com | identity_org1, identity_org2 | 0 | external | identity@oath.com
+| 5 | iOS | github:git.ouroath.com | iOS_org1, iOS_org2 | 1 | external | ios@oath.com
+
+### Cache server to store active *buildClusters* in memory when the service boot up. 
+
+### Below listed apis need to be built to manage the cluster details
+
+| Method | url | Description
+| --- | --- | ---
+| `POST` | ` /buildClusters ` | ` { "name":"iOS", "scmContext":"github:git.ouroath.com", "scmOrganizations": "iOS_org1", "isActive":1, "managedBy": "screwdriver" } `
+| `GET` | `	/buildClusters ` | ` get list of buildClusters info `
+| `GET` | `	/buildClusters/:name ` | ` get a particular buildCluster info `
+| `DELETE` | ` /buildClusters/:name ` | ` delete buildCluster `
+
+
+### Cluster on-board
+
+1. Build cluster admin requesting access with cluster info and user credentials
+2. buildCluster table populated with cluster info
+3. SD admin to create queue based on #1
+4. SD admin authorize build cluster user and queue 
+
+
+## Flow
+### SD Validator
+
+	1. SD validator should validate if the annotated buildCluster in yaml is onboarded and active. 
+	
+### Screwdriver API to Scheduler service
+
+	1. Screwdriver UI / PR commit / Merge triggers new build via Screwdriver API
+	2. Screwdriver API inturn calls Scheduler service with appropriate build details to schedule a job
+	3. Authentication via JWT token (refer Authorization section)
+	4. Scheduler service queries `buildClusters` cache for active records with cluster name from build info and validate if job can be scheduled in appropriate buildCluster queue
+	5. one (or) more record exist, then assign job to the queue identified by generating a random number within given boundaries, which is the returned list size of records
+	6. no records, then query `clusters` table for active records with managedBy=screwdriver
+	7. repeat step #4
+	8. Update build info with cluster and queue details
+
+### Queue worker to Scheduler service  
```

Comment:
This is more like Queue worker to Queue, not scheduler, it doesn't need to talk to scheduler

Reply:
@minz1027 ack, will change